managed/services/backup/removal_service.go

```diff
@@ -27,142 +31,300 @@ import (
+// DeleteArtifact deletes specified artifact along with files if specified.
+func (s *RemovalService) DeleteArtifact(storage Storage, artifactID string, removeFiles bool) error {
+	artifact, _, err := s.lockArtifact(artifactID, models.DeletingBackupStatus)
 	if err != nil {
 		return err
 	}
 
-	if s3Config != nil && removeFiles {
-		if err := s.s3.RemoveRecursive(
-			ctx,
-			s3Config.Endpoint,
-			s3Config.AccessKey,
-			s3Config.SecretKey,
-			s3Config.BucketName,
-			// Recursive listing finds all the objects with the specified prefix.
-			// There could be a problem e.g. when we have artifacts `backup-daily` and `backup-daily-1`, so
-			// listing by prefix `backup-daily` gives us both artifacts.
-			// To avoid such a situation we need to append a slash.
-			artifactName+"/"); err != nil {
-			if _, updateErr := models.UpdateArtifact(s.db.Querier, artifactID, models.UpdateArtifactParams{
-				Status: models.BackupStatusPointer(models.FailedToDeleteBackupStatus),
-			}); updateErr != nil {
-				s.l.WithError(updateErr).
-					Errorf("failed to set status %q for artifact %q", models.FailedToDeleteBackupStatus, artifactID)
-			}
+	defer func() {
+		if err != nil {
+			s.setArtifactStatus(artifactID, models.FailedToDeleteBackupStatus)
+		}
+	}()
+
+	restoreItems, err := models.FindRestoreHistoryItems(s.db.Querier, models.RestoreHistoryItemFilters{
+		ArtifactID: artifactID,
+	})
+	if err != nil {
+		return err
+	}
 
+	for _, ri := range restoreItems {
+		if err = models.RemoveRestoreHistoryItem(s.db.Querier, ri.ID); err != nil {
 			return err
 		}
 	}
 
-	return s.db.InTransaction(func(tx *reform.TX) error {
-		restoreItems, err := models.FindRestoreHistoryItems(tx.Querier, models.RestoreHistoryItemFilters{
-			ArtifactID: artifactID,
-		})
+	if removeFiles {
+		go func() {
+			var err error
+			defer func() {
+				if err != nil {
+					s.setArtifactStatus(artifactID, models.FailedToDeleteBackupStatus)
+				}
+			}()
+
+			location, err := models.FindBackupLocationByID(s.db.Querier, artifact.LocationID)
+			if err != nil {
+				s.l.WithError(err).Error("couldn't get location")
+				return
+			}
+
+			service, err := models.FindServiceByID(s.db.Querier, artifact.ServiceID)
+			if err != nil {
+				s.l.WithError(err).Error("couldn't get service")
+				return
+			}
+
+			if err = s.deleteArtifactFiles(context.Background(), storage, location, artifact, len(artifact.MetadataList)); err != nil {
+				s.l.WithError(err).Error("couldn't delete artifact files")
+				return
+			}
+
+			if service.ServiceType == models.MongoDBServiceType && artifact.Mode == models.PITR {
+				if err = s.deleteArtifactPITRChunks(context.Background(), storage, location, artifact, nil); err != nil {
+					s.l.WithError(err).Error("couldn't delete artifact PITR chunks")
+					return
+				}
+			}
+
+			err = models.DeleteArtifact(s.db.Querier, artifactID)
+			if err != nil {
+				s.l.WithError(err).Error("couldn't delete artifact")
+			}
```

Comment:
I think it makes sense to extract it to normal function

Reply:
The func is straightful without loops or something, and used only in one place. Not sure the extacting can add readability. Let's ask other reviewer. @idoqo @atymchuk

Reply:
I'm fine with either way.